pkg/servicecache/service_cache.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package servicecache
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"sync"
+
+	pb "github.com/cilium/hubble/api/v1/flow"
+
+	"github.com/cilium/cilium/api/v1/models"
+)
+
+// ServiceCache is a cache of existing services.
+type ServiceCache struct {
+	mu    sync.RWMutex
+	cache map[string]*entry
```

Comment:
nit: Have you considered using two maps, e.g.:

```go
// index by numerical id
servicesByID map[int64]*entry
// index by frontend "<ip>:<port>"
servicesByAddr map[string]*entry
```

It would save you at the `genIDKey` function.

Reply:
My initial implementation was using 2 maps but it turns out this was turning the process of writing tests into an even more painful experience as you always have to assert equality of two maps instead of just one within the tests. For the negligible overhead that the `genIDKey` function brings, I thought that it was a good trade-off to use only one map instead.

Reply:
Sounds good to me!